src/Agent.Sdk/SecretMasking/OssSecretMasker.cs

```diff
@@ -81,11 +106,132 @@ public void Dispose()
     {
         _secretMasker?.Dispose();
         _secretMasker = null;
+        _telemetry = null;
     }
 
     public string MaskSecrets(string input)
     {
-        return _secretMasker.MaskSecrets(input);
+        _telemetry?.ProcessInput(input);
+        return _secretMasker.MaskSecrets(input, _telemetry?.DetectionAction);
+    }
+
+    public void EnableTelemetry()
+    {
+        _telemetry ??= new Telemetry();
+    }
+
+    /// <summary>
+    /// If enabled via <see cref="TelemetryEnabled"/> opt-in, publishes
+    /// telemetry via the provided callback that handles sending it over the
+    /// wire.
+    ///
+    /// Always publishes a 'SecretMasker' event with overall stats, and may
+    /// publish 'SecretMaskerDetections' events mapping C3ID (12 byte
+    /// non-reversible seeded hash) to pattern moniker ('name'.'id') for
+    /// rule-based patterns with high entropy if any such detections are found.
+    /// </summary>
+    public void PublishTelemetry(PublishSecretMaskerTelemetryAction publishAction)
```

Comment:
>PublishTelemetry [](http://example.com/codeflow?start=16&length=16)

Just noting in this implementation users can publish collected data repeatedly, passing a different publish action for each iteration.

Reply:
This is no longer possible. It is now called StopAndPublishTelemetry, opposite StartTelemetry. And the threading issues are resolved by capitalizing on this.